### ISSUE TITLE
Czar102 - `CAP_PER_ADDRESS` not enforced

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-07-beam-auction-judging#75]
Czar102

medium

# `CAP_PER_ADDRESS` not enforced

## Summary
The cap per address for minting tokens is not enforced – an address may effectively mint more tokens. Hence, core protocol functionality is impaired. It cannot be fixed by redeploying, since the tokens have already been minted in the ERC721 contract.

## Vulnerability Detail
Because anyone can create as many identities on EVM chains, the cap per address can be easily bypassed.

For example, in case of a contract, it may deploy proxies that buy `CAP_PER_ADDRESS` tokens each and then send those to the owner contract and then selfdestruct. The state change is effectively the same as if the core contract bought more than `CAP_PER_ADDRESS` tokens.

## Impact
The cap per address is not enforced. Identities may effectively mint more tokens.

## Code Snippet
```js
        // Cannot mint more than the cap per address
        require(mintedPerAddressBefore + amount <= CAP_PER_ADDRESS, "Mint cap per address reached");
```

## Tool used

Manual Review

## Recommendation
To limit the number of tokens that can be minted by someone, another identity than an address needs to be used. Or, alternatively, one can use a whitelist for determining addresses that can mint up to [`CAP_PER_ADDRESS`](https://github.com/sherlock-audit/2023-07-beam-auction/blob/main/dutch-nft/src/MeritDutchAuction.sol#L12) tokens.

Duplicate of #59

**Comment by Czar102** [comment]
escalate

tl;dr: I don't think it is a duplicate of #59 as the response to that issue from the judge doesn't match the vulnerability described here. ("The state change is effectively the same as if the core contract bought more than `CAP_PER_ADDRESS` tokens.")

To understand this issue's core, let's first consider another hypothetical situation – a simple lending protocol.

Users can lend, repay and transfer their positions when the other party agrees to claim it.

Because the liquidation mechanism is simple – the whole account needs to be liquidated at once, the team makes a design decision to limit the total amount of debt of a single account in a borrow function to a certain threshold in order to allow smaller liquidators (not only entities with large liquidity) to liquidate all loans.

Unfortunately, they forgot to do that in the debt transfer functionality, so two colluding addresses in a permissionless protocol could transfer debt one to the other. Now, a single address can have a lot of debt and smaller liquidators cannot liquidate them.

I think a similar situation happens here – the design choices are usually out of scope - auditors aren't interested how large liquidators need to be in order to liquidate a position, it's up to the protocol team to define that. But since they made this check, it means that it is an important property that the contract should hold, hence any state change that the check forbids should be in general forbidden. If it's not the case, then the bypass method constitutes an exploit.

> But wait, kiddo. A single address can still mint up to `CAP_PER_ADDRESS` tokens, right? It is as the spec said.

The spec said so, yes. But, as previously, it doesn't really matter what operations are performed, if the state change change can be the same as a forbidden state transition, then it is an issue. This means that the spec had a logic error that auditors caught. In the above example, the team specified that only the borrow function has debt limiting capability.


I would also like to note that the team's justification "Cap does not need to be bullet proof and was only intended to slightly inconvenience people wanting to mint multiple" effectively means that the check wasn't supposed to prevent anything. This means that they are now trying to push this impact out of scope, which is not allowed by Sherlock rules because the lack of the effectiveness of this check was not put under the "list of any known issues/acceptable risks that should not result in a valid finding" in the contest README.

The judge's commentary "Technically there is also no bypass, because the the value is a cap per address, not cap per person." does not reflect the issue described here, because no person-like identity are mentioned in the vulnerability detail, only the equivalence of the ERC721 state change "as if" the forbidden state transition was not. Identities are mentioned only in the impact and mitigation sections as that is most likely what the team wants, not the core issue here.

To my best knowledge, there was no other duplicate of #59 that explicitly noted the state change equivalence in the scope of the ERC721. There were some good quality reports showing the same state change, despite never explicitly mentioning its significance as the core issue here. I am happy to help selecting issues that were implicitly mentioning the state change, if those are deemed valid, too.

**Comment by sherlock-admin2** [comment]
> escalate
> 
> tl;dr: I don't think it is a duplicate of #59 as the response to that issue from the judge doesn't match the vulnerability described here. ("The state change is effectively the same as if the core contract bought more than `CAP_PER_ADDRESS` tokens.")
> 
> To understand this issue's core, let's first consider another hypothetical situation – a simple lending protocol.
> 
> Users can lend, repay and transfer their positions when the other party agrees to claim it.
> 
> Because the liquidation mechanism is simple – the whole account needs to be liquidated at once, the team makes a design decision to limit the total amount of debt of a single account in a borrow function to a certain threshold in order to allow smaller liquidators (not only entities with large liquidity) to liquidate all loans.
> 
> Unfortunately, they forgot to do that in the debt transfer functionality, so two colluding addresses in a permissionless protocol could transfer debt one to the other. Now, a single address can have a lot of debt and smaller liquidators cannot liquidate them.
> 
> I think a similar situation happens here – the design choices are usually out of scope - auditors aren't interested how large liquidators need to be in order to liquidate a position, it's up to the protocol team to define that. But since they made this check, it means that it is an important property that the contract should hold, hence any state change that the check forbids should be in general forbidden. If it's not the case, then the bypass method constitutes an exploit.
> 
> > But wait, kiddo. A single address can still mint up to `CAP_PER_ADDRESS` tokens, right? It is as the spec said.
> 
> The spec said so, yes. But, as previously, it doesn't really matter what operations are performed, if the state change change can be the same as a forbidden state transition, then it is an issue. This means that the spec had a logic error that auditors caught. In the above example, the team specified that only the borrow function has debt limiting capability.
> 
> 
> I would also like to note that the team's justification "Cap does not need to be bullet proof and was only intended to slightly inconvenience people wanting to mint multiple" effectively means that the check wasn't supposed to prevent anything. This means that they are now trying to push this impact out of scope, which is not allowed by Sherlock rules because the lack of the effectiveness of this check was not put under the "list of any known issues/acceptable risks that should not result in a valid finding" in the contest README.
> 
> The judge's commentary "Technically there is also no bypass, because the the value is a cap per address, not cap per person." does not reflect the issue described here, because no person-like identity are mentioned in the vulnerability detail, only the equivalence of the ERC721 state change "as if" the forbidden state transition was not. Identities are mentioned only in the impact and mitigation sections as that is most likely what the team wants, not the core issue here.
> 
> To my best knowledge, there was no other duplicate of #59 that explicitly noted the state change equivalence in the scope of the ERC721. There were some good quality reports showing the same state change, despite never explicitly mentioning its significance as the core issue here. I am happy to help selecting issues that were implicitly mentioning the state change, if those are deemed valid, too.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by jkoppel** [comment]
Short of requiring KYC, it is not actually possible to prevent this kind of thing. And even then it can be circumvented. 

There is some ambiguity in the spec, but the interpretation Czar102 argues for is unimplementable. That's why I (and I suspect many others) did not submit this.

**Comment by Czar102** [comment]
Thanks for the comment @jkoppel!

Firstly, I want to note that the goal of the audit is not only to assist with fixing the vulnerabilities, but also determining them. In my experience, it happened a few times that I found a design error in the audit. There was no way to fix, but there usually is a way, even though not obvious, that one can mitigate the issue to some extent. It still may be not bulletproof, but that's up to the protocol team – auditors should share their concerns no matter if the issue is fixable.

Secondly, if you (and potentially others) didn't perceive this as a medium severity vulnerability, it means that they may not have understood the bug in its full extent or lacked the way to think about it in terms that make it a medium severity finding. I hope that reading my escalation explains what I mean here.

Lastly, I think that the protocol team has some rough understanding of what they wanted to achieve by this check, no matter if this is a limit per person, per active address before the auction, or per users that signed up for beta access to some set of projects. It is important to let them know that they need to specify this buyer group more formally if they want the check to be enforced. Otherwise, they can simply remove it from the smart contracts, which will save gas. Please note that, as mentioned previously, using a whitelist doesn't imply KYC at all. It is only if one wants to enforce a per-person limit.

If this issue was not reported, users may have thought, through some communication in the community, that there was an enforced limit per address. The role of the audit is to formalize this meaning if a user wanted to investigate the soundness of the check.

I hope I managed to justify my reasoning.

**Comment by Oot2k** [comment]
From my perspective this issue is a duplicate of #59 (see sherlock duplication rule).

I think laying the focus on the identity part does not make sense, because as previously stated, its not possible to enforce.

But, I believe other explanations on this issue make more sense, provided in the other escalations.

I still think that Watsons did not have the possibility to know that protocol team will set this to a high value.
We are talking about a static/immutable value.

Also there is clear intend to limit the mint, this issue different to other issues because here it leads to loss of funds.

In current contract state the impact of this issue is to be seen, if we ignore comment from sponsor and deploy contracts as they are (which should be the approche of this audit), 
we can conclude that a mint per address cap of 4 is much less then 10.000 (total supply), this means this check has some importance and bypassing it is not expected behavior.  

As previously stated it still remains up to discussion if atomic bypass is bypass enough when talking about mint to address. 
(atomic bypass as bypass in one tx)

I wont comment on this issue anymore, but on others which address same issue that are also escalated.
I think escalation is invalid, this is a duplicate of #59

**Comment by Czar102** [comment]
@Oot2k I stand corrected. What I meant by "is not a duplicate" was that the impact was shown in another way. I hope this mistake in wording does not make the crux of the escalation invalid. What I meant was that this issue should be of medium severity.

**Comment by hrishibhat** [comment]
Result:
Low
Duplicate of #59  
https://github.com/sherlock-audit/2023-07-beam-auction-judging/issues/19#issuecomment-1653662281

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [Czar102](https://github.com/sherlock-audit/2023-07-beam-auction-judging/issues/75/#issuecomment-1648209557): rejected
